### PR TITLE
Get the correct library loading functions at runtime

### PR DESCRIPTION
**Pull request by filmor**
### What does this implement/fix? Explain your changes.

Moves the library loader to three separate classes that contain implementations for Darwin, Linux and Windows s.t. they can be initialized at runtime.

### Does this close any currently open issues?

Nope, but it brings us further towards a single build.

### CODE REVIEW

File: src/runtime/platform/LibraryLoader.cs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                    return new LinuxLoader();
+                default:
+                    throw new Exception($"This operating system ({os}) is not supported");
+            }
+        }
+    }
+
+    class LinuxLoader : ILibraryLoader
+    {
+        private static int RTLD_NOW = 0x2;
```
Comment (lostmsu) [inline src/runtime/platform/LibraryLoader.cs]: const?

File: src/runtime/platform/LibraryLoader.cs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                dllHandle = RTLD_DEFAULT;
+            }
+
+            // clear previous errors if any
+            dlerror();
+            IntPtr res = dlsym(dllHandle, name);
+            IntPtr errPtr = dlerror();
+            if (errPtr != IntPtr.Zero)
+            {
+                throw new Exception("dlsym: " + Marshal.PtrToStringAnsi(errPtr));
```
Comment (lostmsu) [inline src/runtime/platform/LibraryLoader.cs]: Here and in similar places should throw more specific exception.

File: src/runtime/runtime.cs [hunk 3]
```diff
@@ -7,96 +7,7 @@
...
-        public static extern IntPtr LoadLibrary(string dllToLoad);
-
-        [DllImport(NativeDll)]
-        public static extern IntPtr GetProcAddress(IntPtr hModule, string procedureName);
-
-        [DllImport(NativeDll)]
-        public static extern bool FreeLibrary(IntPtr hModule);
-#endif
-    }
+    using Python.Runtime.Platform;
```
Comment (lostmsu) [inline src/runtime/runtime.cs]: Looks like should be outside. Or let's move them all inside in a separate commit.

File: src/runtime/platform/LibraryLoader.cs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            switch (os)
+            {
+                case OperatingSystemType.Windows:
+                    return new WindowsLoader();
+                case OperatingSystemType.Darwin:
+                    return new DarwinLoader();
+                case OperatingSystemType.Linux:
+                    return new LinuxLoader();
+                default:
+                    throw new Exception($"This operating system ({os}) is not supported");
```
Comment (lostmsu) [inline src/runtime/platform/LibraryLoader.cs]: PlatformNotSupportedException

File: src/runtime/platform/LibraryLoader.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    class LinuxLoader : ILibraryLoader
+    {
+        private static int RTLD_NOW = 0x2;
+        private static int RTLD_GLOBAL = 0x100;
+        private static IntPtr RTLD_DEFAULT = IntPtr.Zero;
+        private const string NativeDll = "libdl.so";
+
+        public IntPtr Load(string fileName)
+        {
+            return dlopen($"lib{fileName}.so", RTLD_NOW | RTLD_GLOBAL);
```
Comment (lostmsu) [inline src/runtime/platform/LibraryLoader.cs]: Add error handling

File: src/runtime/platform/LibraryLoader.cs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    class DarwinLoader : ILibraryLoader
+    {
+        private static int RTLD_NOW = 0x2;
+        private static int RTLD_GLOBAL = 0x8;
+        private const string NativeDll = "/usr/lib/libSystem.dylib";
+        private static IntPtr RTLD_DEFAULT = new IntPtr(-2);
+
+        public IntPtr Load(string fileName)
+        {
+            return dlopen($"lib{fileName}.dylib", RTLD_NOW | RTLD_GLOBAL);
```
Comment (lostmsu) [inline src/runtime/platform/LibraryLoader.cs]: Add error handling

File: src/runtime/platform/LibraryLoader.cs [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    class WindowsLoader : ILibraryLoader
+    {
+        private const string NativeDll = "kernel32.dll";
+
+        [DllImport(NativeDll)]
+        static extern IntPtr LoadLibrary(string dllToLoad);
+
+        public IntPtr Load(string dllToLoad) => WindowsLoader.LoadLibrary(dllToLoad);
```
Comment (lostmsu) [inline src/runtime/platform/LibraryLoader.cs]: Add error handling

File: src/runtime/platform/LibraryLoader.cs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+        [DllImport(NativeDll)]
+        static extern IntPtr GetProcAddress(IntPtr hModule, string procedureName);
+
+        public IntPtr GetFunction(IntPtr hModule, string procedureName) => WindowsLoader.GetProcAddress(hModule, procedureName);
```
Comment (lostmsu) [inline src/runtime/platform/LibraryLoader.cs]: Add error handling

**Comment by lostmsu** [comment]
Done with a review pass.

File: src/runtime/platform/LibraryLoader.cs (hunk 2 above)
Comment (lostmsu) [reply src/runtime/platform/LibraryLoader.cs]: Also, Ansi? I'd double-check.

File: src/runtime/platform/LibraryLoader.cs (hunk 1 above)
Comment (filmor) [reply src/runtime/platform/LibraryLoader.cs]: I think `static` is fine, `const` not being available at runtime has created problems in the past and makes debugging more annoying.

File: src/runtime/platform/LibraryLoader.cs (hunk 6 above)
Comment (filmor) [reply src/runtime/platform/LibraryLoader.cs]: Done.

File: src/runtime/platform/LibraryLoader.cs (hunk 4 above)
Comment (filmor) [reply src/runtime/platform/LibraryLoader.cs]: Done.

File: src/runtime/runtime.cs (hunk 3 above)
Comment (filmor) [reply src/runtime/runtime.cs]: Done.

File: src/runtime/platform/LibraryLoader.cs (hunk 2 above)
Comment (filmor) [reply src/runtime/platform/LibraryLoader.cs]: Well, it's expected to be ASCII only, so this should be fine. Also, this whole PR is just a refactoring, it should be functionally equivalent to the previous state.
